spec/api/api_bookmarks_spec.rb

```diff
@@ -0,0 +1,132 @@
+require 'spec_helper'
+require 'api/api_helper'
+
+describe "API BookmarksController" do
+  include ApiHelper
+
+  # Override is_archivist so all users are archivists from this point on
+  class User < ActiveRecord::Base
+    def is_archivist?
+      true
+    end
+  end
+
+  bookmark = { id: "123",
+               url: "http://foo.com",
+               author: "Thing",
+               title: "Title Thing",
+               summary: "<p>blah blah blah</p>",
+               fandom_string: "Testing",
+               rating_string: "General Audiences",
+               category_string: ["M/M"],
+               relationship_string: "Starsky/Hutch",
+               character_string: "Starsky,hutch",
+               notes: "<p>Notes</p>",
+               tag_string: "youpi",
+               collection_names: "",
+               private: "0",
+               rec: "0" }
+
+  describe "Valid API bookmark import" do
+    before do
+      mock_external
+      @user = create(:user)
+    end
+
+    after do
+      WebMock.reset!
+    end
+
+    it "should return 200 OK when all bookmarks are created" do
+      post "/api/v1/bookmarks/import",
+           { archivist: @user.login,
+             bookmarks: [ bookmark ]
+           }.to_json,
+           valid_headers
+      assert_equal 200, response.status
+    end
+
+    it "should return 200 OK when no bookmarks are created" do
+      post "/api/v1/bookmarks/import",
+           { archivist: @user.login,
+             bookmarks: [ bookmark ]
+           }.to_json,
+           valid_headers
+      assert_equal 200, response.status
+    end
+
+    it "should return 200 OK when only some bookmarks are created" do
+      post "/api/v1/bookmarks/import",
+           { archivist: @user.login,
+             bookmarks: [ bookmark, bookmark ]
+           }.to_json,
+           valid_headers
+      assert_equal 200, response.status
+    end
+
+    it "should create bookmarks associated with the archivist" do
+      pseud_id = @user.default_pseud.id
+      post "/api/v1/bookmarks/import",
+           { archivist: @user.login,
+             bookmarks: [ bookmark, bookmark ]
+           }.to_json,
+           valid_headers
+      bookmarks = Bookmark.find_all_by_pseud_id(pseud_id)
+      assert_equal bookmarks.count, 2
```

Comment:
Good point about the "some" bookmarks spec. When I originally wrote the works controller, I returned different HTTP responses for these different states (all, none or some created). But as they're application errors, not HTTP errors, I now think 200 is the code to use. Anyway, I copy-pasted those into the bookmarks spec and forgot to clean up. I'll remove that test altogether.

The second one does expect duplicates to be created because there doesn't seem to be a de-duping mechanism in the Bookmark model. If I bookmark the same work multiple times, I get multiple bookmarks (even though only one is listed under the work, interestingly). This may indeed be a unintended flaw in the Bookmark model, but it feels as if fixing it is beyond the scope of this particular ticket which aims to reproduce the web behaviour in the API (though a separate one which fixes the model and this test sounds like a good idea!)